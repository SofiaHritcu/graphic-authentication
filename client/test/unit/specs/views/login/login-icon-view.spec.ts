import { shallowMount } from "@vue/test-utils";
import "regenerator-runtime/runtime";
import LoginIconView from "@/views/login/login-icon-view.vue";
import mockStore from "../../store/mocked-store-modules";
import { iconsCategoriesMock } from "../../store/modules/icons/mock-data/mock";

describe("Login Icon View", () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = shallowMount(LoginIconView, {
      store: mockStore,
      data: {
        userNameRules: [],
        iconPassCount: 6,
        iconPasses: [],
        iconPassLastCompletedIndex: 0,
        loginFormRef: "loginFormRef",
      },
      computed: {
        onLargerViewPort: () => true,
      },
      mocks: {
        $router: {
          push: jest.fn(),
        },
        $vuetify: {
          breakpoint: {
            name: "",
          },
        },
      },
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  describe("snapshot", () => {
    it("should match snapshot", () => {
      expect(wrapper.vm.$el).toMatchSnapshot();
    });
  });

  describe("data rules", () => {
    describe("userNameRules", () => {
      it("should return a message if the userName is not defined", () => {
        var userNameRules = wrapper.vm.userNameRules;
        expect(userNameRules[0]("")).toBe("username is required");
      });
      it("should return a message if the userName is longer than 10 characters", () => {
        var userNameRules = wrapper.vm.userNameRules;
        expect(userNameRules[1]("abcdefghijklm")).toBe(
          "username must be less than 10 characters"
        );
      });
    });
  });

  describe("methods", () => {
    describe("handleBackBtnClick", () => {
      it("should call router push with the corresponding parameters", async () => {
        expect.assertions(1);
        const spyRouterPush = jest.spyOn(wrapper.vm.$router, "push");

        await wrapper.vm.handleBackBtnClick();

        expect(spyRouterPush).toHaveBeenCalledWith({ name: "ga-login" });
      });
    });
    describe("handleIconCategoryChange", () => {
      it("should make the logo invisible and set the current icons to the corresponding value", async () => {
        expect.assertions(2);

        await wrapper.setData({
          logoVisible: true,
        });

        wrapper.vm.handleIconCategoryChange(iconsCategoriesMock[0].category);

        expect(wrapper.vm.logoVisible).toBe(false);
        expect(wrapper.vm.iconItems).toEqual(iconsCategoriesMock[0].icons);
      });
    });
    describe("handleIconClick", () => {
      it("should make the logo invisible and update the icon passes if the pass is not long enough", async () => {
        expect.assertions(2);

        await wrapper.setData({
          logoVisible: true,
        });
        const iconItem = iconsCategoriesMock[0].icons[0];

        wrapper.vm.handleIconClick(iconItem);

        expect(wrapper.vm.iconPasses).toEqual({
          "0": iconItem,
          "1": "",
          "2": "",
          "3": "",
          "4": "",
          "5": "",
        });
        expect(wrapper.vm.iconPassLastCompletedIndex).toBe(1);
      });
    });
    it("should check the password strength if the pass is complete", async () => {
      expect.assertions(3);

      await wrapper.setData({
        logoVisible: true,
        iconPassLastCompletedIndex: 5,
        iconPasses: {
          "0": "a",
          "1": "a",
          "2": "a",
          "3": "a",
          "4": "a",
          "5": "",
        },
      });
      const iconItem = iconsCategoriesMock[0].icons[0];

      wrapper.vm.handleIconClick(iconItem);

      expect(wrapper.vm.iconPasses).toEqual({
        "0": "a",
        "1": "a",
        "2": "a",
        "3": "a",
        "4": "a",
        "5": iconItem,
      });
      expect(wrapper.vm.iconPassLastCompletedIndex).toBe(6);
      expect(wrapper.vm.iconPassStrengthInterval).toStrictEqual({
        lower: 0,
        upper: 100,
        qualifying: "bad level of security",
        color: "#C62828",
        vColor: "red",
        vOpacity: "darken-3",
      });
    });
    describe("handleClearBtnClick", () => {
      it("should reset all data props related to the pass", () => {
        const spySetUpDefaultIconsPass = jest.spyOn(
          wrapper.vm,
          "setUpDefaultIconsPass"
        );
        wrapper.vm.handleClearBtnClick();

        expect(wrapper.vm.userName).toBe("");
        expect(wrapper.vm.iconPasses).toStrictEqual(["", "", "", "", "", ""]);
        expect(wrapper.vm.iconPassLastCompletedIndex).toBe(0);
        expect(spySetUpDefaultIconsPass).toHaveBeenCalled();
      });
    });
    describe("submitLogin", () => {
      it("should call fetchLogin with the user input", async () => {
        expect.assertions(1);

        const spyOnFetchLogin = jest.spyOn(wrapper.vm, "fetchLogin");
        await wrapper.setData({
          userName: "test",
          passAfterTransformation: "test",
        });

        await wrapper.vm.submitLogin();

        expect(spyOnFetchLogin).toHaveBeenCalledWith({
          userName: "test",
          password: "test",
        });
      });
    });
    describe("handleLoginBtnClick", () => {
      it("should submit login if the form is valid", () => {
        wrapper.vm.$refs.loginFormRef.validate = jest.fn(() => true);
        const spySubmitLogin = jest.spyOn(wrapper.vm, "submitLogin");

        wrapper.vm.handleLoginBtnClick();

        expect(spySubmitLogin).toHaveBeenCalled();
      });
      it("should reset the form if it is not valid", () => {
        wrapper.vm.$refs.loginFormRef.validate = jest.fn(() => false);
        wrapper.vm.$refs.loginFormRef.reset = jest.fn();
        const spyFormReset = jest.spyOn(
          wrapper.vm.$refs.loginFormRef,
          "reset"
        );

        wrapper.vm.handleLoginBtnClick();

        expect(wrapper.vm.loginErrorShows).toBe(true);
        expect(spyFormReset).toHaveBeenCalled();
      });
    });
  });

  describe("computed", () => {
    describe("iconsScrollSize", () => {
      it("should return 400 when being on larger viewports", () => {
        wrapper.vm.$vuetify.breakpoint.name = "lg";
        expect(wrapper.vm.iconsScrollSize).toEqual(400);
      });
      it("should return 200 when being on smaller viewports", () => {
        wrapper.vm.$vuetify.breakpoint.name = "lg";
        expect(wrapper.vm.iconsScrollSize).toEqual(400);
      });
    });
    describe("onLargerViewPort", () => {
      it("should return true for lg or xl breakpoint", () => {
        wrapper.vm.$vuetify.breakpoint.name = "lg";
        expect(wrapper.vm.onLargerViewPort).toEqual(true);
      });
    });
  });
});
