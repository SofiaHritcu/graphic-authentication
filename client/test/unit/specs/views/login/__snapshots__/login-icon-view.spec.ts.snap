// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Icon View snapshot should match snapshot 1`] = `
<div
  class="ga__login_icon_view"
>
  <v-row
    class="ga__login_icon_view__title_row"
  >
    <v-btn
      color="teal darken-2"
      outlined=""
    >
      
      Go Back To Choose IPASS
      
      <fa-icon
        class="ml-2"
        color="#00796B"
        icon="fa-solid fa-arrow-rotate-left"
        size="lg"
      />
    </v-btn>
  </v-row>
   
  <v-row
    align-content="center"
    class="ga__login_icon_view__title"
    justify="center"
  >
    <v-col
      cols="12"
    >
      <div
        class="text-overline white--text text--darken-1 ga__login_icon_view__title__text"
      >
        
        Login
      
      </div>
    </v-col>
  </v-row>
   
  <v-row
    class="ga__login_icon_view__content"
  >
    <v-col
      cols="12"
      lg="6"
    >
      <v-form>
        <v-row
          class="ga__login_icon_view__form_row"
          justify="center"
        >
          <v-col
            class="ga__login_icon_view__form_col"
            cols="8"
          >
            <v-text-field
              class="ga__login_icon_view__form_field"
              color="teal"
              counter="10"
              dark=""
              filled=""
              label="user name"
              required=""
              rules="function (v) {
          /* istanbul ignore next */
          cov_tnpmhnuvo().f[2]++;
          cov_tnpmhnuvo().s[15]++;
          return (
            /* istanbul ignore next */
            (cov_tnpmhnuvo().b[0][0]++, !!v) ||
            /* istanbul ignore next */
            (cov_tnpmhnuvo().b[0][1]++, \\"username is required\\")
          );
        },function (v) {
          /* istanbul ignore next */
          cov_tnpmhnuvo().f[3]++;
          cov_tnpmhnuvo().s[16]++;
          return (
            /* istanbul ignore next */
            (cov_tnpmhnuvo().b[1][0]++, v.length <= 10) ||
            /* istanbul ignore next */
            (cov_tnpmhnuvo().b[1][1]++, \\"username must be less than 10 characters\\")
          );
        }"
            />
          </v-col>
        </v-row>
         
        <v-row
          class="ga__login_icon_view__form_row"
          justify="center"
        >
          <v-col
            class="ga__login_icon_view__form_col"
            cols="8"
          >
            <v-select
              color="teal"
              dark=""
              filled=""
              hint="Select the icons category in order to build up your pass"
              item-color="teal"
              items="alphabet,animals,astronomy,automative,buildings"
              label="icon category"
              persistent-hint=""
              required=""
              value="alphabet"
            />
          </v-col>
        </v-row>
         
        <v-row
          class="hidden"
        >
          <v-col
            class="ga__login_icon_view__icons_title"
            cols="12"
          >
            <div
              class="grey--text text--darken-1"
            >
              
              Choose icons for your
              
              <span
                class="text-overline font-weight-light white--text ga__login_icon_view__icons_title__pass"
              >
                IPASS
              </span>
            </div>
             
            <!---->
             
            <!---->
          </v-col>
        </v-row>
         
        <v-row
          class="ga__login_icon_view__icons_row hidden"
        >
          <v-col
            class="ga__login_icon_view__icon"
            cols="2"
          >
            <div
              class="ga__login_icon_view__icon__content"
            >
              <!---->
            </div>
          </v-col>
          <v-col
            class="ga__login_icon_view__icon"
            cols="2"
          >
            <div
              class="ga__login_icon_view__icon__content"
            >
              <!---->
            </div>
          </v-col>
          <v-col
            class="ga__login_icon_view__icon"
            cols="2"
          >
            <div
              class="ga__login_icon_view__icon__content"
            >
              <!---->
            </div>
          </v-col>
          <v-col
            class="ga__login_icon_view__icon"
            cols="2"
          >
            <div
              class="ga__login_icon_view__icon__content"
            >
              <!---->
            </div>
          </v-col>
          <v-col
            class="ga__login_icon_view__icon"
            cols="2"
          >
            <div
              class="ga__login_icon_view__icon__content"
            >
              <!---->
            </div>
          </v-col>
          <v-col
            class="ga__login_icon_view__icon"
            cols="2"
          >
            <div
              class="ga__login_icon_view__icon__content"
            >
              <!---->
            </div>
          </v-col>
        </v-row>
      </v-form>
    </v-col>
     
    <v-col
      class="ga__login_icon_view__icons"
      cols="12"
      lg="6"
    >
      <v-row
        class="ga__login_icon_view__icons__row"
      >
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__login_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-a"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__login_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-b"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__login_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-c"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__login_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-d"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__login_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-e"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__login_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-f"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__login_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-g"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__login_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-h"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__login_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-i"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__login_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-j"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__login_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-k"
              size="2x"
            />
          </div>
        </v-col>
        <v-col
          align-self="center"
          cols="3"
        >
          <div
            class="ga__login_icon_view__icon_wrapper"
          >
            <fa-icon
              color="white"
              icon="fa-solid fa-l"
              size="2x"
            />
          </div>
        </v-col>
      </v-row>
    </v-col>
  </v-row>
   
  <div
    class="ga__login_icon_view__logo_container"
  >
    <div
      class="ga__login_icon_view__logo"
    />
  </div>
   
  <v-snackbar>
    
    Invalid username!
    
  </v-snackbar>
</div>
`;
